### PR TITLE
Make mesh option-parsable

### PR DESCRIPTION
**Pull request by nikwit**
## Proposed changes

mesh can be options parsed. This will be needed to make ObserveFields options parsable to include an interpolating mesh

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (wthrowe) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp:55]: Question: Do we want to support different extents in different dimensions?

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (wthrowe) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp]: Call this just `Basis`, and then you can remove the `name` function.

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (wthrowe) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp]: Call this just `Quadrature`, and then you can remove the `name` function.

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (wthrowe) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp]: Make sure there are spaces at the end of each line of the string.

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (wthrowe) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp]: Remove the `\\f$`.  That's interpreted by doxygen in some places, but not here.

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (wthrowe) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp]: Drop the last sentence (or at least the part starting from "and can be").  It's not useful to someone creating a Mesh in an input file.

File: src/NumericalAlgorithms/Spectral/Spectral.hpp
Comment (wthrowe) [inline src/NumericalAlgorithms/Spectral/Spectral.hpp]: Why remove the name?

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: Move this declaration inside the for loops, and then you don't need to clear it after each use.

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: Use a `size_t`, and then you don't need a cast when you use it later.

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: Make these a `const std::array<Spectral::Basis, 3>` and a `const std::array<Spectral::Quadratures, 2>`.

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: After the changes to the arrays I suggested above, these can be changed to range-based for loops (`for (const auto basis : bases)`, etc.).  The bases and quadratures can be directly streamed below to get the names into the creation string.

File: src/NumericalAlgorithms/Spectral/Spectral.hpp
Comment (wthrowe) [inline src/NumericalAlgorithms/Spectral/Spectral.hpp:573]: Add a test for this new functionality to `Test_Spectral.cpp`.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Welcome to the project!

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (nikwit) [reply src/NumericalAlgorithms/Spectral/Mesh.hpp:55]: @nilsleiffischer suggested to wait until #2625 is finished and it can be added later.

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (nikwit) [reply tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: I still need the index to get the corresponding enum type though. Could use `std::array<std::pair<std::string, Spectral::Basis>>` with structured bindings and a range-based loop though Im not sure thats easier to read.

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (nilsvu) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp:14]: Please also add the `Options` library to the `CMakeLists.txt` file in this directory, specifically after line 52 (i.e. in the `PUBLIC` section of `target_link_libraries`.

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (nilsvu) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp:232]: My clang-format also suggests this formatting, but `git-clang-format` doesn't. I think one of them reads the `.clang-format` file in the repo root and the other doesn't (maybe). That's as far as I have looked into this, but it sure annoys me occasionally.

I suggest you run `git-clang-format develop` on this branch. Formatting-only changes are annoying to review because they obscure real changes. If want to fix existing formatting-issues in a file I recommend you interactively rebase to jump ahead of your commits, run clang-format over the file and make a separate commit. Such formatting-only commits are super easy to review and don't distract from the real diff :)

If you do this, please force-push the formatting edits directly instead of pushing them as fixup commits to keep the git history in the PR manageable.

File: src/NumericalAlgorithms/Spectral/Spectral.cpp
Comment (nilsvu) [inline src/NumericalAlgorithms/Spectral/Spectral.cpp:627]: delete blank line

File: src/NumericalAlgorithms/Spectral/Spectral.cpp
Comment (nilsvu) [inline src/NumericalAlgorithms/Spectral/Spectral.cpp]: delete comma

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (nilsvu) [reply tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: With Will's suggestion you would just create the mesh as `Mesh<Dim>{num_points, basis, quadrature}`

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (nilsvu) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp:248]: Would it make sense to template this on `size_t Dim` so you can test meshes of any dim?

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (nilsvu) [reply tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: And add `Quadrature::FaceCentered` and `Quadrature::CellCentered` please. We'll actually use those because commonly we want to interpolate to a uniform grid for visualization.

**Review by nilsvu — CHANGES_REQUESTED** [review]
Since this is your first PR I'd like to point you to our contributing guide, which explains our pull-request workflow to you:

https://spectre-code.org/contributing_to_spectre.html#pull-request-reviews

In particular, please push changes you make in response to reviews as commits that are prefixed with the word `fixup`. This helps reviewers because they don't have to re-review the entire diff :) We'll let you know once we have had a chance to look at the fixups so you can squash them together and rebase on latest `develop`. Just ping me on Slack if anything is unclear.

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (wthrowe) [reply src/NumericalAlgorithms/Spectral/Mesh.hpp:55]: OK.  Sounds good.

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/Mesh.hpp:12]: - The commit message should be active: `Make Mesh option createable`
- There should be 2 commits. One for making `Spectral::Basis` option createable and one for making `Mesh` option createable. @nilsleiffischer can probably help you with splitting a commit in a video chat :)

File: src/NumericalAlgorithms/Spectral/Spectral.cpp
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/Spectral.cpp]: why not just extend the `cond` from above?

File: src/NumericalAlgorithms/Spectral/Spectral.cpp
Comment (nilsdeppe) [reply src/NumericalAlgorithms/Spectral/Spectral.cpp]: No, ` Chebyshev, Legendre, or FiniteDifference`, so "replace first ` or` with `,` :)

File: src/NumericalAlgorithms/Spectral/Mesh.hpp
Comment (nilsvu) [reply src/NumericalAlgorithms/Spectral/Mesh.hpp:12]: No worries @nilsdeppe we're dealing with a professional here ;)

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (nikwit) [reply tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: my bad, I didnt see you changed the type of the array.

**Review by nilsvu — COMMENTED** [review]
LGMT, please squash once @wthrowe and @nilsdeppe have had a chance to look

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: Leave this one as a traditional for loop, not range-based.

Question for the general audience: Should Mesh be checking that its extents are valid (i.e., within the allowed range for the basis/quadrature)?  I notice it is accepting extents of zero here.

**Review by wthrowe — CHANGES_REQUESTED** [review]
You can squash once the third reviewer agrees.  You can also directly squash in the change I request below without showing a fixup commit first.

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp
Comment (nilsvu) [reply tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: I think that would require we implement a compile-time-to-runtime switch for the min and max num collocation points in `Spectral`. Certainly sounds useful.

Comment (nilsdeppe) [reply tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.cpp]: That seems like a good thing to check :) I think currently the check is delayed until we get the matrices for whatever linear operator we are using, but checking earlier seems much better. I think in this case it's the difference between a parse error and a runtime error, with a parse error being preferred.

**Comment by nilsvu** [comment]
@nilsdeppe could you take a quick look at the fixup?

**Review by nilsdeppe — COMMENTED** [review]
LGTM. Nice work @nikwit!

No need to wait for me to approve explicitly :) 

I also think it's reasonable to defer the check in `Mesh` that the extents is valid for the basis to a future PR, but I'll let the other reviewers decide on that :)

**Comment by wthrowe** [comment]
Oh, yes, not for this PR.  Just wanted to check there isn't a reason we allow out-of-range values.

**Comment by nilsvu** [comment]
@nikwit I'm afraid you accidentally squashed the formatting commit together with the "Make Basis option creatable" commit. Could you fix that before we merge this?

**Comment by nilsvu** [comment]
Actually, nvm, if @wthrowe is OK with this I'm happy to merge as-is since it's all reviewed and good to go.

**Review by nilsvu — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]
Yeah, not worth worrying about.

**Comment by nilsvu** [comment]
Created https://github.com/sxs-collaboration/spectre/issues/2652 for the extents validation

**Comment by nilsvu** [comment]
Gratulations on your first contribution @nikwit! Please feel free to add your name to the list of `Contributors` in the `Metadata.yaml` file and submit a PR for that change. This means your name will appear on the citable SpECTRE DOI at https://zenodo.org/record/4290405 for the next release.

**Pull request merged**
